File: app/src/main/java/com/winlator/cmod/ContainersFragment.java
package com.winlator.cmod;

import android.content.Context;
import android.content.Intent;
import android.graphics.Rect;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ContextThemeWrapper;
import androidx.appcompat.widget.PopupMenu;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.winlator.cmod.container.Container;
import com.winlator.cmod.container.ContainerManager;
import com.winlator.cmod.container.Shortcut;
import com.winlator.cmod.contentdialog.ContainerSettingsComposeDialog;
import com.winlator.cmod.contentdialog.ContentDialog;
import com.winlator.cmod.contentdialog.StorageInfoDialog;
import com.winlator.cmod.contents.ContentsManager;
import com.winlator.cmod.core.PreloaderDialog;
import com.winlator.cmod.google.ContainerBackupManager;
import com.winlator.cmod.xenvironment.ImageFs;

import java.util.ArrayList;
import java.util.List;

public class ContainersFragment extends Fragment {
    private static final int VIEW_TYPE_ADD = 0;
    private static final int VIEW_TYPE_CONTAINER = 1;

    private RecyclerView recyclerView;
    private TextView emptyTextView;
    private ContainerManager manager;
    private PreloaderDialog preloaderDialog;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preloaderDialog = new PreloaderDialog(getActivity());
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (getActivity() instanceof AppCompatActivity) {
            androidx.appcompat.app.ActionBar actionBar =
                    ((AppCompatActivity) getActivity()).getSupportActionBar();
            if (actionBar != null) {
                actionBar.setTitle(R.string.common_ui_containers);
            }
        }
        manager = new ContainerManager(getContext());
        loadContainersList();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.containers_fragment, container, false);
        recyclerView = view.findViewById(R.id.RecyclerView);
        emptyTextView = view.findViewById(R.id.TVEmptyText);

        GridLayoutManager gridLayoutManager = new GridLayoutManager(getContext(), 3);
        recyclerView.setLayoutManager(gridLayoutManager);

        int spacing = (int) (8 * getResources().getDisplayMetrics().density);
        recyclerView.addItemDecoration(new RecyclerView.ItemDecoration() {
            @Override
            public void getItemOffsets(@NonNull Rect outRect, @NonNull View v,
                                       @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
                int position = parent.getChildAdapterPosition(v);
                int column = position % 3;
                outRect.left = column * spacing / 3;
                outRect.right = spacing - (column + 1) * spacing / 3;
                outRect.bottom = spacing;
            }
        });

        return view;
    }

    private void openAddContainer() {
        if (!ImageFs.find(getContext()).isValid()) {
            Toast.makeText(getContext(), R.string.setup_wizard_system_image_not_installed, Toast.LENGTH_LONG).show();
            return;
        }
        new Thread(() -> {
            Context ctx = getContext();
            if (ctx == null) return;
            boolean installed = ContentsManager.hasInstalledRuntimes(ctx);
            if (!isAdded()) return;
            requireActivity().runOnUiThread(() -> {
                if (!isAdded()) return;
                if (!installed) {
                    Toast.makeText(getContext(), R.string.container_no_wine_installed, Toast.LENGTH_LONG).show();
                    return;
                }
                new ContainerSettingsComposeDialog(requireActivity(), null, this::loadContainersList).show();
            });
        }).start();
    }

    private void loadContainersList() {
        ArrayList<Container> containers = manager.getContainers();
        recyclerView.setAdapter(new ContainersAdapter(containers));
        emptyTextView.setVisibility(containers.isEmpty() ? View.VISIBLE : View.GONE);
    }

    private void showContainerBackupsDialog(Container container) {
        ContentDialog dialog = new ContentDialog(requireContext());
        dialog.setTitle(R.string.container_backups_title);
        dialog.setMessage(R.string.container_backups_prompt);

        TextView backupButton = dialog.findViewById(R.id.BTConfirm);
        TextView restoreButton = dialog.findViewById(R.id.BTCancel);
        backupButton.setText(R.string.google_cloud_backup);
        restoreButton.setText(R.string.google_cloud_restore);

        dialog.setOnConfirmCallback(() -> startContainerBackup(container));
        dialog.setOnCancelCallback(() -> startContainerRestore(container));
        dialog.show();
    }

    private void startContainerBackup(Container container) {
        preloaderDialog.show(R.string.container_backups_backing_up);
        ContainerBackupManager.backupContainer(requireActivity(), container, result -> {
            preloaderDialog.close();
            ContentDialog.alert(requireContext(), result.message, () -> {});
        });
    }

    private void startContainerRestore(Container container) {
        preloaderDialog.show(R.string.container_backups_checking);
        ContainerBackupManager.prepareRestore(requireActivity(), container, preparation -> {
            if (!isAdded()) {
                preloaderDialog.close();
                return;
            }

            if (!preparation.success) {
                preloaderDialog.close();
                ContentDialog.alert(requireContext(), preparation.message, () -> {});
                return;
            }

            if (preparation.matchedFile != null) {
                executeContainerRestore(container, preparation.matchedFile);
                return;
            }

            preloaderDialog.close();
            showBackupSelectionDialog(container, preparation.candidates);
        });
    }

    private void showBackupSelectionDialog(Container container, List<ContainerBackupManager.DriveBackupFile> backups) {
        if (backups == null || backups.isEmpty()) {
            ContentDialog.alert(requireContext(), R.string.container_backups_no_files, () -> {});
            return;
        }

        String[] names = new String[backups.size()];
        for (int i = 0; i < backups.size(); i++) {
            names[i] = backups.get(i).name;
        }

        new AlertDialog.Builder(requireContext())
                .setTitle(R.string.container_backups_select_title)
                .setItems(names, (dialog, which) -> executeContainerRestore(container, backups.get(which)))
                .setNegativeButton(R.string.common_ui_cancel, null)
                .show();
    }

    private void executeContainerRestore(Container container, ContainerBackupManager.DriveBackupFile driveFile) {
        preloaderDialog.show(R.string.container_backups_restoring);
        ContainerBackupManager.restoreContainerFromDriveFile(requireActivity(), container, driveFile, result -> {
            preloaderDialog.close();
            ContentDialog.alert(requireContext(), result.message, () -> {});
        });
    }

    private class ContainersAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
        private final List<Container> data;

        private class AddViewHolder extends RecyclerView.ViewHolder {
            private AddViewHolder(View view) {
                super(view);
                view.setOnClickListener(v -> openAddContainer());
            }
        }

        private class ContainerViewHolder extends RecyclerView.ViewHolder {
            private final ImageView runButton;
            private final ImageView editButton;
            private final ImageView duplicateButton;
            private final ImageView menuButton;
            private final ImageView imageView;
            private final TextView title;

            private ContainerViewHolder(View view) {
                super(view);
                this.runButton = view.findViewById(R.id.BTRun);
                this.editButton = view.findViewById(R.id.BTEdit);
                this.duplicateButton = view.findViewById(R.id.BTDuplicate);
                this.imageView = view.findViewById(R.id.ImageView);
                this.title = view.findViewById(R.id.TVTitle);
                this.menuButton = view.findViewById(R.id.BTMenu);
            }
        }

        public ContainersAdapter(List<Container> data) {
            this.data = data;
        }

        @Override
        public int getItemViewType(int position) {
            return position == 0 ? VIEW_TYPE_ADD : VIEW_TYPE_CONTAINER;
        }

        @Override
        public final int getItemCount() {
            return data.size() + 1; // +1 for the add card
        }

        @NonNull
        @Override
        public final RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            if (viewType == VIEW_TYPE_ADD) {
                View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.container_add_card, parent, false);
                return new AddViewHolder(view);
            } else {
                View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.container_list_item, parent, false);
                return new ContainerViewHolder(view);
            }
        }

        @Override
        public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
            if (holder instanceof ContainerViewHolder) {
                ContainerViewHolder ch = (ContainerViewHolder) holder;
                ch.runButton.setOnClickListener(null);
                ch.editButton.setOnClickListener(null);
                ch.duplicateButton.setOnClickListener(null);
                ch.menuButton.setOnClickListener(null);
            }
            super.onViewRecycled(holder);
        }

        @Override
        public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, int position) {
            if (holder instanceof ContainerViewHolder) {
                ContainerViewHolder ch = (ContainerViewHolder) holder;
                final Container item = data.get(position - 1); // offset by 1 for add card
                ch.imageView.setImageResource(R.drawable.ic_containers);
                ch.title.setText(item.getName());

                ch.runButton.setOnClickListener(view -> runContainer(item));
                ch.editButton.setOnClickListener(view -> editContainer(item));
                ch.duplicateButton.setOnClickListener(view -> duplicateContainer(item));
                ch.menuButton.setOnClickListener(view -> showListItemMenu(view, item));
            }
        }

        private void runContainer(Container container) {
            final Context context = getContext();
            Intent intent = new Intent(context, XServerDisplayActivity.class);
            intent.putExtra("container_id", container.id);
            requireActivity().startActivity(intent);
        }

        private void editContainer(Container container) {
            new ContainerSettingsComposeDialog(requireActivity(), container,
                    ContainersFragment.this::loadContainersList).show();
        }

        private void duplicateContainer(Container container) {
            ContentDialog.confirm(getContext(), R.string.containers_list_confirm_duplicate, () -> {
                preloaderDialog.show(R.string.containers_list_duplicating, false);
                manager.duplicateContainerAsync(container, progress -> {
                    preloaderDialog.setProgress(progress);
                }, () -> {
                    preloaderDialog.setProgress(100);
                    preloaderDialog.closeWithDelay(600);
                    new android.os.Handler().postDelayed(() -> loadContainersList(), 600);
                });
            });
        }

        private void showListItemMenu(View anchorView, Container container) {
            final Context context = getContext();
            Context popupContext = new ContextThemeWrapper(context, R.style.ThemeOverlay_ContentPopupMenu);
            PopupMenu listItemMenu = new PopupMenu(popupContext, anchorView);
            listItemMenu.getMenuInflater().inflate(R.menu.container_popup_menu, listItemMenu.getMenu());
            listItemMenu.setForceShowIcon(true);

            listItemMenu.setOnMenuItemClickListener((menuItem) -> {
                switch (menuItem.getItemId()) {
                    case R.id.container_duplicate:
                        duplicateContainer(container);
                        break;
                    case R.id.container_backups:
                        showContainerBackupsDialog(container);
                        break;
                    case R.id.container_remove:
                        ContentDialog.confirm(getContext(), R.string.containers_list_confirm_remove, () -> {
                            preloaderDialog.show(R.string.containers_list_removing);
                            for (Shortcut shortcut : manager.loadShortcuts()) {
                                if (shortcut.container == container)
                                    ShortcutsFragment.disableShortcutOnScreen(context, shortcut);
                            }
                            manager.removeContainerAsync(container, () -> {
                                preloaderDialog.close();
                                loadContainersList();
                            });
                        });
                        break;
                    case R.id.container_info:
                        (new StorageInfoDialog(getActivity(), container)).show();
                        break;
                }
                return true;
            });
            listItemMenu.show();
        }
    }
}

File: app/src/main/feature/settings/containers/ContainersFragment.kt
package com.winlator.cmod.feature.settings
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import com.winlator.cmod.R
import com.winlator.cmod.feature.settings.ContainerSettingsComposeDialog
import com.winlator.cmod.feature.shortcuts.ShortcutsFragment
import com.winlator.cmod.feature.sync.google.ContainerBackupManager
import com.winlator.cmod.runtime.container.Container
import com.winlator.cmod.runtime.container.ContainerManager
import com.winlator.cmod.runtime.content.ContentsManager
import com.winlator.cmod.runtime.display.XServerDisplayActivity
import com.winlator.cmod.runtime.display.environment.ImageFs
import com.winlator.cmod.shared.android.AppUtils
import com.winlator.cmod.shared.io.FileUtils
import com.winlator.cmod.shared.ui.dialog.PreloaderDialog
import com.winlator.cmod.shared.theme.WinNativeTheme
import java.io.File
import kotlin.math.roundToInt

class ContainersFragment : Fragment() {
    private lateinit var manager: ContainerManager
    private lateinit var preloaderDialog: PreloaderDialog

    private var screenState by mutableStateOf(ContainersScreenState())
    private var pendingBackups = emptyList<ContainerBackupManager.DriveBackupFile>()
    private var storageScanToken = 0L

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        preloaderDialog = PreloaderDialog(requireActivity())
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View =
        ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                WinNativeTheme(
                    colorScheme =
                        darkColorScheme(
                            primary = Color(0xFF1A9FFF),
                            background = Color(0xFF18181D),
                            surface = Color(0xFF1C1C2A),
                        ),
                ) {
                    ContainersScreen(
                        state = screenState,
                        onAddContainer = ::openAddContainer,
                        onRunContainer = ::runContainer,
                        onEditContainer = ::editContainer,
                        onDuplicateContainer = ::duplicateContainer,
                        onShowBackups = ::showContainerBackupsDialog,
                        onRemoveContainer = ::removeContainer,
                        onShowInfo = ::showContainerInfo,
                        onDismissDialog = ::dismissDialog,
                        onConfirmDuplicateDialog = ::performDuplicateContainer,
                        onConfirmRemoveDialog = ::performRemoveContainer,
                        onConfirmBackupDialog = ::startContainerBackup,
                        onConfirmRestoreDialog = ::startContainerRestore,
                        onClearCacheDialog = ::clearContainerCache,
                        onBackupSelectionChosen = ::restoreBackupByIndex,
                    )
                }
            }
        }

    override fun onViewCreated(
        view: View,
        savedInstanceState: Bundle?,
    ) {
        super.onViewCreated(view, savedInstanceState)
        (activity as? AppCompatActivity)?.supportActionBar?.setTitle(R.string.common_ui_containers)
        loadContainersList()
    }

    override fun onResume() {
        super.onResume()
        loadContainersList()
    }

    private fun loadContainersList() {
        val context = context ?: return
        manager = ContainerManager(context)
        screenState = screenState.copy(containers = manager.containers.toList())
    }

    private fun openAddContainer() {
        val context = context ?: return
        if (!ImageFs.find(context).isValid) {
            AppUtils.showToast(context, R.string.setup_wizard_system_image_not_installed, Toast.LENGTH_LONG)
            return
        }

        Thread {
            val ctx = context ?: return@Thread
            val installed = ContentsManager.hasInstalledRuntimes(ctx)
            if (!isAdded) return@Thread

            requireActivity().runOnUiThread {
                if (!isAdded) return@runOnUiThread
                if (!installed) {
                    AppUtils.showToast(ctx, R.string.container_no_wine_installed, Toast.LENGTH_LONG)
                    return@runOnUiThread
                }
                ContainerSettingsComposeDialog(requireActivity(), null, ::loadContainersList).show()
            }
        }.start()
    }

    private fun runContainer(container: Container) {
        val ctx = context ?: return
        startActivity(
            Intent(ctx, XServerDisplayActivity::class.java).apply {
                putExtra("container_id", container.id)
            },
        )
    }

    private fun editContainer(container: Container) {
        ContainerSettingsComposeDialog(requireActivity(), container, ::loadContainersList).show()
    }

    private fun duplicateContainer(container: Container) {
        screenState = screenState.copy(dialog = ContainersDialogUiState.ConfirmDuplicate(container))
    }

    private fun removeContainer(container: Container) {
        screenState = screenState.copy(dialog = ContainersDialogUiState.ConfirmRemove(container))
    }

    private fun showContainerInfo(container: Container) {
        screenState =
            screenState.copy(
                dialog =
                    ContainersDialogUiState.StorageInfo(
                        ContainerStorageInfoUiState(container = container),
                    ),
            )
        startStorageScan(container)
    }

    private fun showContainerBackupsDialog(container: Container) {
        screenState = screenState.copy(dialog = ContainersDialogUiState.Backups(container))
    }

    private fun startContainerBackup(container: Container) {
        dismissDialog()
        preloaderDialog.show(R.string.container_backups_backing_up)
        ContainerBackupManager.backupContainer(requireActivity(), container) { result ->
            preloaderDialog.close()
            screenState =
                screenState.copy(
                    dialog =
                        ContainersDialogUiState.Message(
                            title = getString(R.string.container_backups_title),
                            message = result.message,
                        ),
                )
        }
    }

    private fun startContainerRestore(container: Container) {
        dismissDialog()
        preloaderDialog.show(R.string.container_backups_checking)
        ContainerBackupManager.prepareRestore(requireActivity(), container) { preparation ->
            if (!isAdded) {
                preloaderDialog.close()
                return@prepareRestore
            }

            if (!preparation.success) {
                preloaderDialog.close()
                screenState =
                    screenState.copy(
                        dialog =
                            ContainersDialogUiState.Message(
                                title = getString(R.string.container_backups_title),
                                message = preparation.message,
                            ),
                    )
                return@prepareRestore
            }

            if (preparation.matchedFile != null) {
                executeContainerRestore(container, preparation.matchedFile)
                return@prepareRestore
            }

            preloaderDialog.close()
            showBackupSelectionDialog(container, preparation.candidates)
        }
    }

    private fun showBackupSelectionDialog(
        container: Container,
        backups: List<ContainerBackupManager.DriveBackupFile>?,
    ) {
        if (backups.isNullOrEmpty()) {
            screenState =
                screenState.copy(
                    dialog =
                        ContainersDialogUiState.Message(
                            title = getString(R.string.container_backups_title),
                            message = getString(R.string.container_backups_no_files),
                        ),
                )
            return
        }

        pendingBackups = backups
        screenState =
            screenState.copy(
                dialog =
                    ContainersDialogUiState.BackupSelection(
                        container = container,
                        backupNames = backups.map { it.name },
                    ),
            )
    }

    private fun executeContainerRestore(
        container: Container,
        driveFile: ContainerBackupManager.DriveBackupFile,
    ) {
        dismissDialog()
        preloaderDialog.show(R.string.container_backups_restoring)
        ContainerBackupManager.restoreContainerFromDriveFile(requireActivity(), container, driveFile) { result ->
            preloaderDialog.close()
            screenState =
                screenState.copy(
                    dialog =
                        ContainersDialogUiState.Message(
                            title = getString(R.string.container_backups_title),
                            message = result.message,
                        ),
                )
        }
    }

    private fun performDuplicateContainer(container: Container) {
        dismissDialog()
        preloaderDialog.show(R.string.containers_list_duplicating, false)
        manager.duplicateContainerAsync(container, { progress ->
            preloaderDialog.setProgress(progress)
        }) {
            preloaderDialog.setProgress(100)
            preloaderDialog.closeWithDelay(600)
            Handler(Looper.getMainLooper()).postDelayed({
                loadContainersList()
            }, 600)
        }
    }

    private fun performRemoveContainer(container: Container) {
        val ctx = context ?: return
        dismissDialog()
        preloaderDialog.show(R.string.containers_list_removing)
        for (shortcut in manager.loadShortcuts()) {
            if (shortcut.container == container) {
                ShortcutsFragment.disableShortcutOnScreen(ctx, shortcut)
            }
        }
        manager.removeContainerAsync(container) {
            preloaderDialog.close()
            loadContainersList()
        }
    }

    private fun clearContainerCache(container: Container) {
        val cacheDir = File(container.rootDir, ".cache")
        Thread {
            FileUtils.clear(cacheDir)
            container.putExtra("desktopTheme", null)
            container.saveData()
            if (!isAdded) return@Thread
            Handler(Looper.getMainLooper()).post {
                if (!isAdded) return@post
                showContainerInfo(container)
            }
        }.start()
    }

    private fun restoreBackupByIndex(index: Int) {
        val dialog = screenState.dialog as? ContainersDialogUiState.BackupSelection ?: return
        val backup = pendingBackups.getOrNull(index) ?: return
        executeContainerRestore(dialog.container, backup)
    }

    private fun dismissDialog() {
        storageScanToken += 1L
        screenState = screenState.copy(dialog = ContainersDialogUiState.None)
    }

    private fun startStorageScan(container: Container) {
        val token = System.nanoTime()
        storageScanToken = token
        Thread {
            val driveCBytes = calculateDirectorySize(File(container.rootDir, ".wine/drive_c"))
            val cacheBytes = calculateDirectorySize(File(container.rootDir, ".cache"))
            val totalBytes = driveCBytes + cacheBytes
            val internalStorage = FileUtils.getInternalStorageSize().coerceAtLeast(1L)
            val usedPercent =
                ((totalBytes.toDouble() / internalStorage.toDouble()) * 100.0)
                    .toFloat()
                    .coerceIn(0f, 100f)

            if (!isAdded) return@Thread
            Handler(Looper.getMainLooper()).post {
                if (!isAdded || storageScanToken != token) return@post
                val dialog = screenState.dialog as? ContainersDialogUiState.StorageInfo ?: return@post
                if (dialog.data.container.id != container.id) return@post
                screenState =
                    screenState.copy(
                        dialog =
                            ContainersDialogUiState.StorageInfo(
                                dialog.data.copy(
                                    driveCBytes = driveCBytes,
                                    cacheBytes = cacheBytes,
                                    totalBytes = totalBytes,
                                    usedPercent = usedPercent,
                                    isLoading = false,
                                ),
                            ),
                    )
            }
        }.start()
    }

    private fun calculateDirectorySize(root: File?): Long {
        if (root == null || !root.exists()) return 0L
        if (root.isFile) return root.length()

        var total = 0L
        val stack = ArrayDeque<File>()
        stack.add(root)

        while (stack.isNotEmpty()) {
            val current = stack.removeLast()
            val files = current.listFiles() ?: continue
            for (file in files) {
                if (file.isDirectory) {
                    stack.add(file)
                } else {
                    total += file.length()
                }
            }
        }
        return total
    }
}
